src/libos/crates/host-socket/src/stream/states/connected/send.rs

```diff
@@ -208,14 +204,56 @@ impl<A: Addr + 'static, R: Runtime> ConnectedStream<A, R> {
         inner.io_handle.replace(handle);
     }
 
-    fn cancel_send_requests(&self) {
-        let io_uring = self.common.io_uring();
-        let inner = self.sender.inner.lock().unwrap();
-        if let Some(io_handle) = &inner.io_handle {
-            unsafe { io_uring.cancel(io_handle) };
-        } else {
+    // Cancel blocking send requests if any.
+    // Normally, We will always try to send as long as the kernel send buf is not empty. However, if cancel_send_requests is called and
+    // need_wait is true, that means the socket is closed. We will stop sending in this case.
+    pub async fn cancel_send_requests(&self, need_wait: bool) {
+        {
+            let io_uring = self.common.io_uring();
+            let inner = self.sender.inner.lock().unwrap();
+            if let Some(io_handle) = &inner.io_handle {
+                unsafe { io_uring.cancel(io_handle) };
+            } else {
+                return;
+            }
+        }
+
+        // Wait for the cancel to complete if needed
+        if !need_wait {
             return;
         }
+
+        let poller = Poller::new();
+        let mask = Events::ERR | Events::IN;
+        self.common.pollee().connect_poller(mask, &poller);
+
+        loop {
+            let pending_request_exist = {
+                let inner = self.sender.inner.lock().unwrap();
+                inner.io_handle.is_some()
+            };
+
+            if pending_request_exist {
+                let mut timeout = Some(Duration::from_secs(10));
+                let ret = poller.wait_timeout(timeout.as_mut()).await;
+                if let Err(e) = ret {
+                    warn!("wait cancel send request error = {:?}", e.errno());
+                    // Here we do cancel again. Considering this case, the first cancel does not take effect and the send returns
+                    // successfully and the kernel send buffer is not empty, thus it will send again. Thus, we should submit cancel request again.
+                    // This is only needed for send, because send requests will keep being submitted after shutdown.
+                    let io_uring = self.common.io_uring();
```

Comment:
Consider adding a `ForceShutdown` or `Close` flag to avoid sending cancel requests again.